### ISSUE TITLE
Cucumber Expressions: Non Capture Group

### ISSUE THREAD

**Issue by KyleFairns** [cucumber/cucumber#131]
## Summary

I'm linking to the issue that I created over in the cucumber-js repo [here](https://github.com/cucumber/cucumber-js/issues/784) in order to trace back the whole conversation.

There are many cases in our step definitions where we have _a varity of repeated groups that don't change the steps functionality, but does do quite a bit for readablilty_ in the feature files:

* "must"/"should"
* optional words like "the", "I" etc that would match `Given I fill in "Foo"` & `And fill in "Bar"`
* lists of interchangable words

Say that you are using these alternatives across a large set of step definitions and you wish to change them all. This is where the problem (and the solution) both appear.

## Current way of completing the task

At current, we can do alternatives like so:

```javascript
Given('{pronoun} have/has {int} child(ren)', (pronoun, childCount) => {
  // Code that does not require the "pronoun" parameter at all
});
```

However, this does mean that we have a parameter that is never actually used in the step definition that we have and this is the only way of storing the list to be used in many differing steps.

## Possible Solution

Perhaps square brackets could be used for things that don't need to be captured, but still require the transformation.

```javascript
Given('[pronoun] have/has {int} child(ren)', (childCount) => {
   // Code
});

When('[pronoun] picks up {stringInDoubleQuotes} from {stringInDoubleQuotes}', (child, place) => {
   // Code
});

Then('[pronoun] should have {int} child(ren) in [pronoun] care', (childCount) => {
  // Code
});

```
Another plus side, is that if the parameter needs to be captured, the only difference would be the brackets that you use in that step definition:

```javascript
  Then('{pronoun} does something', (pronoun) => {
     if(pronoun == "Jeff"){
        // Do something different for Jeff
     } else {
       // Do your thing
     }   
  });
```

## Context & Motivation

We will be working with cucumber on a wide variety of projects in our company and I can guarantee that many of them will require a large number of declaratively written step definitions that require many alternatives that won't be captured for use within the step definitions.

This would make it easier for us to keep track of all of the alternatives that we will be using, and will mean that we wont have to search through the many step definition files for all of the uses..

And I am certain that others in the cucumber community will be in the same situation as I.

**Comment by mattwynne** [comment]
Couldn't you just use regular string interpolation here?

i.e.
```
const pronoun = "He/She"
Given(`${pronoun} have/has {int} child(ren)`, (pronoun, childCount) => {
  // Code that does not require the "pronoun" parameter at all
});
```

**Comment by aslakhellesoy** [comment]
In JavaScript and Ruby you could, @matt, but not in Java or C#.

Further, it becomes cumbersome to use this consistently across different stepdef files.

**Comment by KyleFairns** [comment]
I've completed the changes for this, but I don't believe I have the right permissions to complete a pull request.

```Permission to cucumber/cucumber.git denied to KyleFairns.``` appears when I try to push to a new branch

There was only one file changed for the code, and three tests were removed that tested the escaping of square brackets (however, they are now special characters in terms of cucumber-expressions).

Is there a way for me to gain those permissions, or would you like me to just post the changes below?

**Comment by aslakhellesoy** [comment]
You have to fork the repo, clone your fork create a branch for your change, push to your fork and then make a pull request.

See https://help.github.com/articles/creating-a-pull-request/ for details.

**Comment by aslakhellesoy** [comment]
We now have 3 proposed fixes for this issue:

* #179 - has unresolved code review comments
* #216 - fixes the bug with a more elaborate regexp-parsing regexp than #220, but for js only
* #220 - implements a fix for java/javascript/ruby

I'm going to improve #220 with the technique from #216 and merge that.

I realise this implementation doesn't handle nested groups, as explained in comments in #216, but we'll handle that in a different issue.

**Comment by aslakhellesoy** [comment]
Sigh - please disregard previous comment about the 3 fixes - that comment should have been for #220 which is *not* related to this issue.

Leaving this one open

**Comment by aslakhellesoy** [comment]
Closed due to inactivity
